From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: minaripenguin <minaripenguin@users.noreply.github.com>
Date: Mon, 28 Oct 2024 08:01:28 +0800
Subject: [PATCH 087/226] LayoutInflater: Fallback to reflection when view
 tryCreateViewDirect fails

* Attempt to resolve:
05-01 15:16:47.407 24374 24374 E AndroidRuntime: Caused by: java.lang.IndexOutOfBoundsException
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.content.res.StringBlock.nativeGetString(Native Method)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.content.res.StringBlock.getSequence(StringBlock.java:126)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.content.res.ApkAssets.getStringFromPool(ApkAssets.java:358)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.content.res.AssetManager.getPooledStringForCookie(AssetManager.java:878)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.content.res.TypedArray.loadStringValueAt(TypedArray.java:1402)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.content.res.TypedArray.getValueAt(TypedArray.java:1387)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.content.res.TypedArray.getFont(TypedArray.java:1044)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.widget.TextView.readTextAppearance(TextView.java:4319)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.widget.TextView.<init>(TextView.java:1225)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.widget.TextView.<init>(TextView.java:1162)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.widget.TextView.<init>(TextView.java:1158)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.tryCreateViewDirect(LayoutInflater.java:1277)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.createView(LayoutInflater.java:698)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.createView(LayoutInflater.java:664)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at com.android.internal.policy.PhoneLayoutInflater.onCreateView(PhoneLayoutInflater.java:58)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.onCreateView(LayoutInflater.java:826)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.onCreateView(LayoutInflater.java:846)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:900)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:857)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.rInflate(LayoutInflater.java:1019)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:980)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.rInflate(LayoutInflater.java:1022)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:980)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.rInflate(LayoutInflater.java:1022)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:980)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.inflate(LayoutInflater.java:570)
05-01 15:16:47.407 24374 24374 E AndroidRuntime: 	at android.view.LayoutInflater.inflate(LayoutInflater.java:462)

Signed-off-by: minaripenguin <minaripenguin@users.noreply.github.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 core/java/android/view/LayoutInflater.java | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/core/java/android/view/LayoutInflater.java b/core/java/android/view/LayoutInflater.java
index 3bb76b89b0..3a06b96487 100644
--- a/core/java/android/view/LayoutInflater.java
+++ b/core/java/android/view/LayoutInflater.java
@@ -696,7 +696,12 @@ public abstract class LayoutInflater {
             Trace.traceBegin(Trace.TRACE_TAG_VIEW, name);
 
             // Opportunistically create view directly instead of using reflection
-            View view = tryCreateViewDirect(prefixedName, viewContext, attrs);
+            View view;
+            try {
+                view = tryCreateViewDirect(prefixedName, viewContext, attrs);
+            } catch (Exception e) {
+                view = null;
+            }
             if (view == null) {
                 Constructor<? extends View> constructor = sConstructorMap.get(name);
                 if (constructor != null && !verifyClassLoader(constructor)) {
